### PR TITLE
mesa: 24.1.6 -> 24.2.1, now with 100% less Libgbm Thing(tm)

### PR DESCRIPTION
**Pull request by K900**
## Description of changes

Bit too late in the cycle but we can get this in.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/libraries/mesa/common.nix [hunk 1]
```diff
@@ -1,18 +1,18 @@
-{ lib, fetchurl }:
+{ lib, fetchFromGitLab }:
 # When updating this package, please verify at least these build (assuming x86_64-linux):
 # nix build .#mesa .#pkgsi686Linux.mesa .#pkgsCross.aarch64-multiplatform.mesa .#pkgsMusl.mesa
 # Ideally also verify:
 # nix build .#legacyPackages.x86_64-darwin.mesa .#legacyPackages.aarch64-darwin.mesa
```
Comment (emilazy) [inline pkgs/development/libraries/mesa/common.nix:5]: Should these go in `passthru.tests`?

Comment (K900) [reply pkgs/development/libraries/mesa/common.nix:5]: We can't really do that without instantiating a whole bunch of extra nixpkgses.

Comment (emilazy) [reply pkgs/development/libraries/mesa/common.nix:5]: If you took `pkgsi686Linux` etc. as arguments they’d still only be instantiated if you evaluated `passthru.tests`, wouldn’t they?

Comment (K900) [reply pkgs/development/libraries/mesa/common.nix:5]: Yeah, ofborg/hydra mostly.

**Review by PedroHLC — COMMENTED** [review]
Diff LGTM, when possible I'm going to start to test this (patch and updated outputs) with my user base.

But I think this patch deserves some explanation on “what it does?”, “why?”, and “how it achieves that?”. (A short paragraph should be enough)

**Comment by K900** [comment]
Updated to 24.2.0-rc4, also updated the libgbm patch to severely simplify the loading logic.

**Comment by Atemu** [comment]
xwayland fails to build:

```
[447/447] Linking target hw/xwayland/Xwayland
FAILED: hw/xwayland/Xwayland 
gcc  -o hw/xwayland/Xwayland hw/xwayland/Xwayland.p/meson-generated_relative-pointer-unstable-v1-protocol.c.o hw/xwayland/Xwayland.p/meson-generated_pointer-constraints-unstable-v1-protocol.c.o hw/xwayland/Xwayland.p/meson-generated_pointer-gestures-unstable-v1-protocol.c.o hw/xwayland/Xwayland.p/meson-generated_tablet-unstable-v2-protocol.c.o hw/xwayland/Xwayland.p/meson-generated_xwayland-keyboard-grab-unstable-v1-protocol.c.o hw/xwayland/Xwayland.p/meson-generated_xdg-output-unstable-v1-protocol.c.o hw/xwayland/Xwayland.p/meson-generated_linux-dmabuf-unstable-v1-protocol.c.o hw/xwayland/Xwayland.p/meson-generated_viewporter-protocol.c.o hw/xwayland/Xwayland.p/meson-generated_xdg-shell-protocol.c.o hw/xwayland/Xwayland.p/meson-generated_drm-lease-v1-protocol.c.o hw/xwayland/Xwayland.p/meson-generated_keyboard-shortcuts-inhibit-unstable-v1-protocol.c.o hw/xwayland/Xwayland.p/meson-generated_xwayland-shell-v1-protocol.c.o hw/xwayland/Xwayland.p/meson-generated_tearing-control-v1-protocol.c.o hw/xwayland/Xwayland.p/meson-generated_fractional-scale-v1-protocol.c.o hw/xwayland/Xwayland.p/meson-generated_linux-drm-syncobj-v1-protocol.c.o hw/xwayland/Xwayland.p/meson-generated_drm-protocol.c.o hw/xwayland/Xwayland.p/xwayland.c.o hw/xwayland/Xwayland.p/xwayland-input.c.o hw/xwayland/Xwayland.p/xwayland-cursor.c.o hw/xwayland/Xwayland.p/xwayland-drm-lease.c.o hw/xwayland/Xwayland.p/xwayland-pixmap.c.o hw/xwayland/Xwayland.p/xwayland-present.c.o hw/xwayland/Xwayland.p/xwayland-screen.c.o hw/xwayland/Xwayland.p/xwayland-shm.c.o hw/xwayland/Xwayland.p/xwayland-output.c.o hw/xwayland/Xwayland.p/xwayland-cvt.c.o hw/xwayland/Xwayland.p/xwayland-vidmode.c.o hw/xwayland/Xwayland.p/xwayland-window.c.o hw/xwayland/Xwayland.p/xwayland-window-buffers.c.o hw/xwayland/Xwayland.p/.._.._mi_miinitext.c.o hw/xwayland/Xwayland.p/xwayland-xtest.c.o hw/xwayland/Xwayland.p/xwayland-glamor.c.o hw/xwayland/Xwayland.p/xwayland-dmabuf.c.o hw/xwayland/Xwayland.p/xwayland-glamor-gbm.c.o hw/xwayland/Xwayland.p/xwayland-glamor-xv.c.o -Wl,--as-needed -Wl,--no-undefined -fvisibility=hidden -Wl,--start-group dix/liblibxserver_main.a glamor/libglamor.a fb/liblibxserver_fb.a mi/liblibxserver_mi.a dix/liblibxserver_dix.a composite/liblibxserver_composite.a damageext/liblibxserver_damageext.a dbe/liblibxserver_dbe.a randr/liblibxserver_randr.a miext/damage/liblibxserver_miext_damage.a render/liblibxserver_render.a present/liblibxserver_present.a Xext/liblibxserver_xext.a miext/sync/liblibxserver_miext_sync.a xfixes/liblibxserver_xfixes.a Xi/liblibxserver_xi.a xkb/liblibxserver_xkb.a record/liblibxserver_record.a os/liblibxserver_os.a os/liblibxlibc.a dri3/liblibxserver_dri3.a Xext/liblibxserver_xext_vidmode.a xkb/liblibxserver_xkb_stubs.a Xi/liblibxserver_xi_stubs.a glx/liblibxserver_glx.a glx/liblibglxvnd.a /nix/store/0zmji8sw3jm1j4akllcwc6g7i8cilyza-systemd-256.2/lib/libsystemd.so /nix/store/65jrlvybibj5hc1dg93p540ajqp9spnp-pixman-0.43.4/lib/libpixman-1.so /nix/store/53ra2xc82ayzh80bxf4m2iw5479aiy4v-libxkbfile-1.1.3/lib/libxkbfile.so /nix/store/apdf947wc488gp3xrw3clx89ll742abc-libXfont2-2.0.6/lib/libXfont2.so /nix/store/5y3sa1zg076nfgr0ihp71ia2ykk1160y-libXdmcp-1.1.5/lib/libXdmcp.so /nix/store/wkz8w1jhaw53lc4h3190lvnwnvn3jm7l-libunwind-1.8.1/lib/libunwind.so /nix/store/hgvky83ijhl0fl625g4hwllwzmpdnksl-wayland-1.23.0/lib/libwayland-client.so /nix/store/3chy0r06q15a725gbvk3bcv00fmqrrwd-libxcvt-0.1.2/lib/libxcvt.so /nix/store/wclfrpl95p9wbcn1ff0ybfr718vs06cf-libxshmfence-1.3.2/lib/libxshmfence.so /nix/store/v5ni0z9an6gb57ync6zhhdvaypsip6p3-libdrm-2.4.122/lib/libdrm.so /nix/store/viwa56vxk41gsyrwwk90m86ra1awgxhp-libepoxy-1.5.10/lib/libepoxy.so /nix/store/2n6lnbblbad633dly84ign6dhrkc4z6l-libei-1.2.1/lib/libei.so /nix/store/2n6lnbblbad633dly84ign6dhrkc4z6l-libei-1.2.1/lib/liboeffis.so /nix/store/fyl50v9gk6n47r8gf4qah4bam1975k6v-mesa-24.2.0-rc4/lib/libgbm.so /nix/store/5mlcqyhcnv3nqn76kyakqcwac7g8qbxz-libdecor-0.2.2/lib/libdecor-0.so -lm -ldl -lcrypto /nix/store/9mp72xkxzr3n1fgl4nhrm4jyxpwkzjcv-libtirpc-1.3.4/lib/libtirpc.so -lpthread /nix/store/r3q5i3lhgn75f3pjh7p43fl6dimhh7wj-libXau-1.0.11/lib/libXau.so /nix/store/rcvwvfgk5dkwc3a8sza5q0351h8axbkh-libglvnd-1.7.0/lib/libGL.so -Wl,--end-group
/nix/store/qsx2xqqm0lp6d8hi86r4y0rz5v9m62wn-binutils-2.42/bin/ld: warning: libgallium-24.2.0-rc4.so, needed by /nix/store/fyl50v9gk6n47r8gf4qah4bam1975k6v-mesa-24.2.0-rc4/lib/libgbm.so, not found (try using -rpath or -rpath-link)
/nix/store/qsx2xqqm0lp6d8hi86r4y0rz5v9m62wn-binutils-2.42/bin/ld: /nix/store/fyl50v9gk6n47r8gf4qah4bam1975k6v-mesa-24.2.0-rc4/lib/libgbm.so: undefined reference to `dri_loader_get_extensions@libgallium-24.2.0-rc4.so'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Comment by PedroHLC** [comment]
1. `24.2.0` is out -- But remembering newcomers/visitors, we always wait **.1
2. @K900 so far, nobody showed up with issues related to the patch changes (I didn't rebase to -rc4 patch though).

**Comment by alyssais** [comment]
Have we talked to upstream about the patch yet?

**Comment by K900** [comment]
So, the libgbm thing has been discussed a bunch on Matrix for a very long time, but I never sat down and wrote up a proper rationale for doing this. Here goes.

First, the original problem statement: libgbm is basically a fancy memory allocator API for your GPU. To allocate memory, it needs to talk to your graphics driver (duh). 

In the case of Mesa, that driver is Mesa itself, which libgbm is also part of, so Mesa assumes (generally correctly) that the libgbm and Mesa on a system will always be from the same build. This is true for conventional distros, but we're, well, not very conventional. We can totally have applications built against an old libgbm running on top of new Mesa, and our users generally don't expect it to explode.

Unfortunately, it does, because the libgbm Mesa ("dri") backend talks to the driver using Mesa internal APIs, and if those no longer line up, kaboom. In fact, upstream noticed the potential kaboom, and [made it so the application just aborts immediately instead of slowly melting](https://gitlab.freedesktop.org/mesa/mesa/-/commit/1026d29344192755dd340d6ac13a9674189d2d61#f691faef5372e94be95ad1bdfde1f6efdbc03ac2_630_636).

This means that the state of the art for "I want to run an application built against another libgbm on my new drivers" is "don't", which is not great.

Now, Mesa 24.2 brings with it another problem: [libgbm now links libgallium_dri](https://gitlab.freedesktop.org/mesa/mesa/-/commit/93511c1c5c5fb60166c806d417e4b4378bf1fb31), which is like 40 megabytes of _stuff_ that needs to end up in `mesa.out`, because that's where libgbm lives. That's not really _fatal_, but still annoying.

So here's the proposed solution: libgbm actually has support for loadable backends, that's used by vendor drivers to provide their own implementations, just like libglvnd, ocl-icd, vulkan-loader etc already do. What if we use the same functionality for Mesa itself? Then libgbm basically becomes a driver _loader_, instead of a driver, and we put the matching `gbm_dri.so` into `/run/opengl-drivers` along with its Mesa, and never have the problem again.

The reason I think this could actually be wanted upstream as well is that anyone running their own local build of Mesa runs into the same issues unless they LD_PRELOAD the right libgbm, which is awkward and comes with the usual LD_WHATEVER caveats. It also makes the loader logic surprisingly simple (and it can be simplified even more, at the cost of making the patch more invasive) - we basically just try a bunch of named backends in order using the exact same code paths for Mesa and non-Mesa. Moving the libgallium_dri linked bits to a separate library also saves 40MB for any distro that wants to ship the drivers separately like we do.

The downside of this is that gbm_dri becomes public API, so it has to follow the usual GBM backend ABI stability constraints, but this doesn't seem to be a problem in practice, as the API has not changed in years.

Also, this does potentially unlock decoupling libgbm from Mesa entirely, which is something no one but NixOS probably cares about, but whatever.

**Comment by K900** [comment]
@ofborg eval

**Comment by K900** [comment]
I've split the commit in half: the Mesa bump itself is actually pretty trivial if we're fine with the 40MB $out increase, and The Libgbm Thing(tm)(r) applies very minimal changes on top of that.

**Comment by Atemu** [comment]
Have you talked with upstream about this yet?

**Comment by K900** [comment]
No, I've had some Real Life happening lately. I'll see if I can reach out this week.

**Comment by K900** [comment]
Split off the libgbm thing into https://github.com/NixOS/nixpkgs/pull/338109, because this cycle has been slow enough to just get this in now that 24.2.1 is out.

File: pkgs/development/libraries/mesa/common.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/libraries/mesa/common.nix:5]: We're doing that in intel-media-driver and so far no one complained https://github.com/NixOS/nixpkgs/blob/4dfe798decc582285d0d4e3ec426db524ca1b10f/pkgs/development/libraries/intel-media-driver/default.nix#L59

**Comment by NickCao** [comment]
Shall we also address https://github.com/NixOS/nixpkgs/issues/328923?

**Comment by K900** [comment]
We should, but there's no rush on that.

**Pull request merged**